[Scheduler] extract schedule context to own object

kbrock opened the pull request:
I'm back in the scheduler and trying to determine which jobs are scheduled.

There is a lot of boiler plate code here.
Removed 1 set of duplication to trim it down.

There is more duplication that could be tackled, but these 2 changes seem like a move in the right direction.

Where I'm going with this:

- default `:job => true` for cron jobs (all jobs to be honest)
- default the category tag
- possibly combine `schedule_enabled?` with `schedule_for`

**update:** removed the dsl like syntax and just made a configurator

/cc @imtayadeway yay
/cc @jrafanie thought you'd be interested. If not, push off to someone else

jrafanie commented on app/models/miq_schedule_worker/runner.rb:
```diff
@@ -59,58 +59,58 @@ def load_system_schedules
+    context == :all || @active_roles.include?(context.to_s)
+  end
+
+  def scheduler_for(context)
```
Maybe role or category instead of context?

jrafanie commented on app/models/miq_schedule_worker/runner.rb:
```diff
@@ -59,58 +59,58 @@ def load_system_schedules
-    @system_scheduler.schedule_every(*args, &block)
-  rescue ArgumentError => err
-    _log.error("#{err.class} for schedule_every with #{args.inspect}.  Called from: #{caller[1]}.")
+  def schedule_enabled?(context)
```
Again, maybe role or category.  What do you think?

kbrock replied:
yes, `role` is better

jrafanie commented on app/models/miq_schedule_worker/runner.rb:
```diff
@@ -59,58 +59,58 @@ def load_system_schedules
     # These schedules need to be run on all servers regardless of the server's role
-    @schedules[:all] ||= []
-
+    s = scheduler_for(:all)
```
I'm not a fan of `s` here when it's scope is more than 10-20 lines.  Maybe `role_scheduler` or even `all_scheduler`.  What do you think?

kbrock replied:
@jrafanie tried `schedule` but maybe `scheduler` may make more sense

jrafanie commented on app/models/miq_schedule_worker/runner.rb:
```diff
@@ -329,20 +324,19 @@ def schedules_for_ems_metrics_coordinator_role
-    @schedules[:event] ||= []
-
+    return unless schedule_enabled?(:event)
+    s = scheduler_for(:event)
```
Again, I"m not a fan of `s` here, it's short but we lose the context of what role we're on here as you page down in each of these methods.

jrafanie commented:
I'm liking the goal of this PR so far.  Nice @kbrock.  I had a few naming/nits but I like it.  @imtayadeway what do you think?

kbrock commented on app/models/miq_schedule_worker/runner.rb:
```diff
@@ -59,58 +59,58 @@ def load_system_schedules
+
+  def scheduler_for(context)
+    @schedules[context] ||= []
+    ::MiqScheduleWorker::MiqScheduleContext.new(self, @schedules[context], @system_scheduler)
```
I tried this. getting

```ruby
MiqScheduleContext.new(self, @schedules[context], @system_scheduler)
# =>
NameError:
        uninitialized constant MiqScheduleWorker::Runner::MiqScheduleContext
```

Ideas how to get around this?

jrafanie replied:
Ugh, yes, I guess this is fine the way it is then.

> Can anything be done to separate that out better?

@kbrock @imtayadeway  we're only adding to the array of schedules, `@schedules`, right?  We can just return the schedules added instead.

kbrock commented:
- changed `context` to `role`
- renamed `MiqScheduleContext` to `MiqScheduler`
- renamed `MiqScheduleContext#schedule_list` to `role_schedules`
- added specs for `MiqScheduler` (thanks @imtayadeway for letting me steal those)
- changed those pesky `s` to `scheduler`

jrafanie commented on app/models/miq_schedule_worker/miq_scheduler.rb:
```diff
@@ -0,0 +1,29 @@
+    # list of schedules for a particular role (i.e.: @schedules[role])
+    attr_accessor :role_schedule
+
+    # MiqScheduleContext.new(self, @schedules[:all], @system_scheduler)
```
@kbrock outdated comment
https://twitter.com/nzkoz/status/538892801941848064

jrafanie commented on app/models/miq_schedule_worker/miq_scheduler.rb:
```diff
@@ -0,0 +1,29 @@
+class MiqScheduleWorker
+  class MiqScheduler
```
We're in the MiqScheduleWorker namespace, it's probably ok to call this Scheduler.

jrafanie commented on app/models/miq_schedule_worker/miq_scheduler.rb:
```diff
@@ -0,0 +1,29 @@
+    def initialize(logger, role_schedule, rufus_scheduler)
+      @logger = logger
+      @role_schedule = role_schedule
+      @rufus_scheduler = rufus_scheduler
```
nit, vertically align `=` here

imtayadeway commented:
@kbrock love the tests! If we can enforce that contract, I think that will give us greater confidence in refactoring more. Thanks!

kbrock commented:
kicked. not sure what was wrong

jrafanie commented:
@imtayadeway I'm good with this,  can you confirm it looks good to you?

@kbrock Is this for euwe?  I assume yes because not including it in the e-release will make it hard to backport anything.

imtayadeway commented:
:+1:

jrafanie commented:
Marking this for euwe.  There's too much chance for backport conflicts considering the amount of changes lines here.

chessbyte commented:
Euwe Backport details:
```sh
$ git log -1
commit 060709335c370ced57c9751196c3d597ea1055e2
Author: Joe Rafaniello <jrafanie@users.noreply.github.com>
Date:   Mon Oct 17 14:39:05 2016 -0400

    Merge pull request #11812 from kbrock/schedule_worker_dsl_2
    
    [Scheduler] extract schedule context to own object
    (cherry picked from commit 397d8744ae13cefddf8efd0e68f6e27824dbbedb)
```
